TrackletAlgorithm/MatchProcessor.h

```diff
@@ -803,51 +820,104 @@ void MatchProcessor(BXType bx,
+      //if the projection goes to the second bin ivmMinus+1 phiProjBin is one
+      ap_uint<1> phiProjBin = 0; 
+
       if (extrabits == ((1U << nextrabits) - 1) && iphi != ((1U << nbits_vmme) - 1)) {
-	shift = 1;
+	//shift = 1;
 	ivmPlus++;
       }
       unsigned int ivmMinus = iphi;
       if (extrabits == 0 && iphi != 0) {
-	shift = -1;
+	phiProjBin = 1; //projection is to next been, but we also search here
+	//shift = -1;
 	ivmMinus--;
       }
       
       ///////////////
       // VMProjection
+      // New code for using new VMSME memory module
+      ap_uint<4> entries_zfirst[8];
+#pragma HLS ARRAY_PARTITION variable=entries_zfirst dim=0 complete
+      ap_uint<4> entries_zlast[8];
+#pragma HLS ARRAY_PARTITION variable=entries_zlast dim=0 complete
+
+      (entries_zfirst[7], entries_zfirst[6], entries_zfirst[5],
+       entries_zfirst[4], entries_zfirst[3], entries_zfirst[2],
+       entries_zfirst[1],entries_zfirst[0]) = 
+	//instubdata.get_mem_entries8A()[bx][zfirst];      
+	instubdata.get_mem_entries8A()[(bx&3)*8+zfirst];      
+
+      (entries_zlast[7], entries_zlast[6], entries_zlast[5],
+       entries_zlast[4], entries_zlast[3], entries_zlast[2],
+       entries_zlast[1],entries_zlast[0]) = 
+	//instubdata.get_mem_entries8B()[bx][zlast];      
+	instubdata.get_mem_entries8B()[(bx&3)*8+zlast];      
+
+      /*
+      std::cout << " z phi: " << zfirst << " " << ivmMinus << " "
+           << nvmstubs[zfirst][ivmMinus]<< " " << entries_zfirst[ivmMinus]
+      << std::endl;
+      */
+
+      ap_uint<1> stubfirstMinus = vmstubsmask[zfirst][ivmMinus];
+      ap_uint<1> stubfirstPlus = vmstubsmask[zfirst][ivmPlus]&&
+                                 (ivmMinus!=ivmPlus);
+      ap_uint<1> stublastMinus = vmstubsmask[zlast][ivmMinus]&&
+                                 (zfirst!=zlast);
+      ap_uint<1> stublastPlus = vmstubsmask[zlast][ivmPlus]&&
+                                 (zfirst!=zlast)&&(ivmMinus!=ivmPlus);;
       
-      ap_uint<4> nstubfirstMinus=nvmstubs[zfirst][ivmMinus];
-      ap_uint<4> nstublastMinus=nvmstubs[zlast][ivmMinus];
-      ap_uint<4> nstubfirstPlus=nvmstubs[zfirst][ivmPlus];
-      ap_uint<4> nstublastPlus=nvmstubs[zlast][ivmPlus];
+      ap_uint<4> nstubfirstMinus = entries_zfirst[ivmMinus];
+      ap_uint<4> nstubfirstPlus = entries_zfirst[ivmPlus];
+      ap_uint<4> nstublastMinus = entries_zlast[ivmMinus];
+      ap_uint<4> nstublastPlus = entries_zlast[ivmPlus];
       
-      if (ivmMinus==ivmPlus) {
-	nstubfirstPlus = 0;
-	nstublastPlus = 0;
-      }
-      if (zfirst==zlast) {
-	nstublastMinus = 0;
-	nstublastPlus = 0;
-      }
 
+
+      //if (ivmMinus==ivmPlus) {
+      //	nstubfirstPlus = 0;
+      //	nstublastPlus = 0;
+      //}
+      //if (zfirst==zlast) {
+      //	nstublastMinus = 0;
+      //	nstublastPlus = 0;
+      //}
+
+      ap_uint<4> maskstubs=(stublastPlus, stubfirstPlus, stublastMinus, stubfirstMinus);
```

Comment:
@jasonfan393 There are a lot of 4s and 8s floating around here, as well as in MemoryTemplateBinnedCM.h. Could you please replace these with named constants, ideally sharing them between here and MemoryTemplateBinnedCM.h if appropriate?

Reply:
@aehart working on replacing magic 4s and 8s now. Just wondering if you have any suggestion on how to handle removing the reference to 8 in line 848. I guess this can be replaced with a loop, but unsure if that will be implemented different by HLS. Curious if there is a way to still do this with the parentheses without explicit reference to 6,5,4 etc.

`
      (entries_zfirst[7], entries_zfirst[6], entries_zfirst[5],
       entries_zfirst[4], entries_zfirst[3], entries_zfirst[2],
       entries_zfirst[1],entries_zfirst[0]) = 
	//instubdata.get_mem_entries8A()[bx][zfirst];      
	instubdata.get_mem_entries8A()[(bx&3)*8+zfirst];      
`
`
      (entries_zlast[7], entries_zlast[6], entries_zlast[5],
       entries_zlast[4], entries_zlast[3], entries_zlast[2],
       entries_zlast[1],entries_zlast[0]) = 
	//instubdata.get_mem_entries8B()[bx][zlast];      
	instubdata.get_mem_entries8B()[(bx&3)*8+zlast];      
`

Reply:
A loop with the unroll pragma might work well here.

Reply:
@aehart -removed most of the 4s and 8s highlighted . There are a lot of them in MemoryTemplateBinnedCM prior to any changes I've made, which will also need changes in TP if I want to rename all these 'nentries8*' things. Should we just put this on a separate PR/issue for now?

Reply:
I think it's fine to have a limited amount of hard-coded numbers, especially in the names of data members, and especially while things are under such intensive development. We can think about how to make it prettier when everything works.